pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +261,291 @@ def simulate(
     tensor([0.68117888, 0.        , 0.31882112, 0.        ], requires_grad=True))
 
     """
+    if circuit.shots and has_mid_circuit_measurements(circuit):
+        return simulate_native_mcm(
+            circuit, rng=rng, prng_key=prng_key, debugger=debugger, interface=interface
+        )
     state, is_state_batched = get_final_state(circuit, debugger=debugger, interface=interface)
     if state_cache is not None:
         state_cache[circuit.hash] = state
     return measure_final_state(circuit, state, is_state_batched, rng=rng, prng_key=prng_key)
+
+
+# pylint: disable=too-many-arguments
+def simulate_native_mcm(
+    circuit: qml.tape.QuantumScript,
+    rng=None,
+    prng_key=None,
+    debugger=None,
+    interface=None,
+) -> Result:
+    """Simulate a single quantum script with native mid-circuit measurements.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. If None, a random key will be
+            generated. Only for simulation using JAX.
+        debugger (_Debugger): The debugger to use
+        interface (str): The machine learning interface to create the initial state with
+        state_cache=None (Optional[dict]): A dictionary mapping the hash of a circuit to the pre-rotated state. Used to pass the state between forward passes and vjp calculations.
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    if circuit.shots.has_partitioned_shots:
+        results = []
+        for s in circuit.shots:
+            aux_circuit = qml.tape.QuantumScript(
+                circuit.operations,
+                circuit.measurements,
+                shots=s,
+                trainable_params=circuit.trainable_params,
+            )
+            results.append(simulate_native_mcm(aux_circuit, rng, prng_key, debugger, interface))
+        return tuple(results)
+    aux_circuit = init_auxiliary_circuit(circuit)
```

Comment:
> it smells of preprocessing that happens at execution time. i'm wondering if somehow this can be refactored into a transform

It seems like transforms would be the way in the current implementation, but I think a tree traversal implementation is in order to make the native mcm evaluation faster. This is unlikely to be expressible in terms of transforms as it is dynamical in nature. So if that's OK, I would merge as is and come back to transforms once we know whether we keep the current implementation or not.

Reply:
I think that's ok, but let's make a backlog story to make sure we come to a solid conclusion on how this code should look

Reply:
[sc-56192]